File: LintCodeJava/BestTimeToBuyAndSellStock.java
public class BestTimeToBuyAndSellStock {
    /**
     * @param prices: Given an integer array
     * @return: Maximum profit
     */

    // V1, O(n), O(1)
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length <= 1) {
            return 0;
        }

        int min = Integer.MAX_VALUE;
        int maxProfit = 0;
        for (int i = 0; i < prices.length; i++) {
            maxProfit = Math.max(maxProfit, prices[i] - min);
            min = Math.min(min, prices[i]);
        }

        return maxProfit;
    }

    // V2, O(n^2)
    // Intuitive
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length <= 1) {
            return 0;
        }

        int maxProfit = 0;
        for (int i = 0; i < prices.length - 1; i++) {
            for (int j = i + 1; j < prices.length; j++) {
                maxProfit = Math.max(maxProfit, prices[j] - prices[i]);
            }
        }

        return maxProfit;
    }
}

File: LeetCode-Python/best_time_to_buy_and_sell_stock.py
# Say you have an array for which the ith element is the price of a given stock on day i.

# If you were only permitted to complete at most one transaction (ie, buy one and sell one share of the stock), design an algorithm to find the maximum profit.

class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit_1(self, prices):
        if len(prices) < 2:
            return 0
        min = prices[0]
        max_diff = 0
        for i in xrange(1, len(prices)):
            if prices[i - 1] < min:
                min = prices[i - 1]
            curr_diff = prices[i] - min
            if curr_diff > max_diff:
                max_diff = curr_diff
        return max_diff